### DISCUSSION TITLE
[Feature Request] Use a keyboard shortcut to trigger a window with content by html

### DISCUSSION THREAD

**Discussion by chengtie** [microsoft/monaco-editor#3654] [Q&A]
### Context

- [X] This issue is not a bug report. *(please use a different template for reporting a bug)*
- [X] This issue is not a duplicate of an existing issue. *(please use the [search](https://github.com/microsoft/monaco-editor/issues) to find existing issues)*

### Description

I would like to add a feature in my editor: users select a text, and click on a keyboard shortcut, then a window with content by html (eg, [a table like this](https://github.com/microsoft/monaco-editor/issues/2856#issuecomment-1001282986)) appears next to the selection, in the end users move the mouse or use another keyboard shortcut to make the window disappear.

I understand that we could use `addDynamicKeybinding` or [like here](https://microsoft.github.io/monaco-editor/playground.html#interacting-with-the-editor-adding-an-action-to-an-editor-instance) to add a keyboard shortcut.

But for the window, it seems that we have only diagnostic window that support html, right? If so, the only use case I know about diagnostic window is by hoverprovider [like here](https://microsoft.github.io/monaco-editor/playground.html#extending-language-services-hover-provider-example), where we need to hover on a text to make the diagnostic window appear.

Does anyone know how to use a keyboard shortcut to trigger a diagnostic window (whose content is related to the selection)?

**Comment by rcjsuen** [comment]
> But for the window, it seems that we have only diagnostic window that support html, right? If so, the only use case I know about diagnostic window is by hoverprovider [like here](https://microsoft.github.io/monaco-editor/playground.html#extending-language-services-hover-provider-example), where we need to hover on a text to make the diagnostic window appear.

You can get the hover to appear with <kbd>Cmd + K, Cmd + I</kbd> on macOS. Not sure what the keybindings are on Windows and Linux but you can figure it out by searching for it with <kbd>F1</kbd>.

![image](https://user-images.githubusercontent.com/15629116/147473807-13154064-12c8-4b48-b144-59b4468a5d30.png)

**Comment by rcjsuen** [comment]
> Now, as far as this question is concerned, I would like to implement another feature. Let's say, show the length (html-supported) of the selected text by a keyboard shortcut. If I still use hoverprovider for this feature, I don't know if it will have conflict with the hinting feature above.
> 
> For instance, I realize that, for a code that has hinting, if I choose only a part of the code, and press `Cmd+K, Cmd+I`, it will still trigger the hinting diagnostic window for the entire code. Ideally, I would want the shortcut to trigger only the display of the length of the selected part of the code.
> 
> Do you have any idea to manage this conflict?

Sorry, I don't think I fully understand. I suggest you create a reproducible piece of code that can be run in the [playground](https://microsoft.github.io/monaco-editor/playground.html) so that people can clearly see what your intent and use case is.

**Comment by chengtie** [comment]
@rcjsuen I deleted my previous comment, and explain my intent here.

I have actually an Excel add-in called Formula Editor. Normally, it is run inside Excel, but we could still get the idea outside: https://v3.10studio.tech/#/formula-editor-addin. You could turn on the toggle and then copy the formula (eg, `=3+2+100+IF(ISERROR(A3:B5),100,A3:B5)`) into the editor.

The editor consists of a Monaco Editor on top of a language server of Excel formulas.

![Screenshot 2021-12-27 at 22 37 55](https://user-images.githubusercontent.com/774409/147509000-a88b2059-0615-4daa-85cb-a7cb83536c69.png)

We have already implemented a hinting feature, that's why `IF(ISERROR(A3:B5),100,A3:B5)` is underlined. When you hover on it, a hinting of rewriting is proposed. This feature was realized by `registerHoverProvider` and `registerCodeActionProvider`. Inside the language server, we have some "non-optimal" patterns (eg, IF+ISERROR), whenever we detect such a pattern in the formula in the editor, we propose a better rewriting (eg, IFERROR).

Now, we would like to implement another feature, it is "evaluation". A user selects a part of the formula (eg, `100`, `A3:B5`, `ISERROR(A3:B5)`, `IF(ISERROR(A3:B5),100,A3:B5)`, `2+100+IF(ISERROR(A3:B5),100,A3:B5)`), press a keyboard shortcut, we will try to evaluate it with the current cell values in worksheets. We want to show the value in form of a table to have a better readability.

So my question is by which APIs we implement this "evaluation" feature.

If we use hoverprovider and `Cmd+K, Cmd+I` as you suggested, I'm afraid there may be conflicts with the hinting feature in some cases. For instance, I realize that when we select `A3:B5` and press `Cmd+K, Cmd+I`, the hinting of rewriting of `IF(ISERROR(A3:B5),100,A3:B5)` appears. But what I want to trigger and show by a shortcut here is the evaluated value of `A3:B5`.

Additionally, I would like this shortcut to be `F9` (`fn+F9` on Mac).

How would you design and implement these features?

**Comment by rcjsuen** [comment]
> How would you design and implement these features?

I don't think what you're describing is easily achievable with the APIs. I imagine you'd have to create a new command that pops up a window yourself.

**Comment by chengtie** [comment]
> > How would you design and implement these features?
> 
> I don't think what you're describing is easily achievable with the APIs. I imagine you'd have to create a new command that pops up a window yourself.

Ah, that's not good news. To create a shortcut that pops up a window that supports html, are there any samples or references to follow?
